test.py

```diff
@@ -0,0 +1,84 @@
+# vstart cluster me running before running this test.
+from subprocess import run as subprocess_run
+from shlex import split as shlex_split
+
+def run_ceph_cmd(cmd, capture_output=False):
+    if isinstance(cmd, str):
+        cmd = shlex_split(cmd)
+    return subprocess_run(args=('./bin/ceph', *cmd),
+                          capture_output=capture_output)
+
+moncap = 'allow rw'
+osdcap = 'allow rw tag cephfs data=a'
+
+fsnamecap = 'fsname=a'
+pathcap = 'path=/dir1'
+rscap = 'root_squash'
+uidcap = 'uid=1000'
+gidscap = 'gids=1000,1001,1002'
+mdscaps = (
+    fsnamecap, pathcap, rscap, uidcap, gidscap,
+
+    f'{fsnamecap} {pathcap}',
+    f'{fsnamecap} {rscap}',
+    f'{fsnamecap} {uidcap}',
+    f'{fsnamecap} {gidscap}',
+    f'{pathcap} {rscap}',
+    f'{pathcap} {uidcap}',
+    f'{pathcap} {gidscap}',
+    f'{rscap} {uidcap}',
+    f'{rscap} {gidscap}',
+    f'{uidcap} {gidscap}',
+
+    f'{fsnamecap} {pathcap} {rscap}',
+    f'{fsnamecap} {pathcap} {uidcap}',
+    f'{fsnamecap} {pathcap} {gidscap}',
+    f'{fsnamecap} {rscap} {uidcap}',
+    f'{fsnamecap} {rscap} {gidscap}',
+    f'{fsnamecap} {uidcap} {gidscap}',
+    f'{pathcap} {rscap} {uidcap}',
+    f'{pathcap} {rscap} {gidscap}',
+    f'{pathcap} {uidcap} {gidscap}',
+    f'{rscap} {uidcap} {gidscap}',
+
+    f'{fsnamecap} {pathcap} {rscap} {uidcap}',
+    f'{fsnamecap} {pathcap} {rscap} {gidscap}',
+    f'{fsnamecap} {pathcap} {uidcap} {gidscap}',
+    f'{fsnamecap} {rscap} {uidcap} {gidscap}',
+    f'{pathcap} {rscap} {uidcap} {gidscap}',
+
+    f'{fsnamecap} {pathcap} {rscap} {uidcap} {gidscap}',
+)
```

Comment:
It's great to systematically try out variations on syntax for a test, and this looks good.

But for checking parsing as this PR is trying to do, there's (at least?) two more important categories of test:

1. Make sure the parsing rejects invalid syntax — otherwise, code that accepts any input can pass!
2. Dump caps and then re-parse them, and make sure they are the same. Similarly, make sure that caps we expect to be equivalent from different parsing input really are.

Reply:
https://github.com/ceph/ceph/pull/51317/commits/087f92b96756121ecfb4fae580066ed47e204449